torchvision/datasets/video_utils.py

```diff
@@ -0,0 +1,190 @@
+    assert tensor.dim() == 1
+    o_stride = tensor.stride(0)
+    numel = tensor.numel()
+    new_stride = (step * o_stride, dilation * o_stride)
+    new_size = ((numel - (dilation * (size - 1) + 1)) // step + 1, size)
+    if new_size[0] < 1:
+        new_size = (0, size)
+    return torch.as_strided(tensor, new_size, new_stride)
+
+
+class VideoClips(object):
+    """
+    Given a list of video files, computes all consecutive subvideos of size
+    `clip_length_in_frames`, where the distance between each subvideo in the
+    same video is defined by `frames_between_clips`.
+    If `frame_rate` is specified, it will also resample all the videos to have
+    the same frame rate, and the clips will refer to this frame rate.
+
+    Creating this instance the first time is time-consuming, as it needs to
+    decode all the videos in `video_paths`. It is recommended that you
+    cache the results after instantiation of the class.
+
+    Recreating the clips for different clip lengths is fast, and can be done
+    with the `compute_clips` method.
+
+    Arguments:
+        video_paths (List[str]): paths to the video files
+        clip_length_in_frames (int): size of a clip in number of frames
+        frames_between_clips (int): step (in frames) between each clip
+        frame_rate (int, optional): if specified, it will resample the video
+            so that it has `frame_rate`, and then the clips will be defined
+            on the resampled video
+    """
+    def __init__(self, video_paths, clip_length_in_frames=16, frames_between_clips=1,
+                 frame_rate=None):
+        self.video_paths = video_paths
+        self._compute_frame_pts()
+        self.compute_clips(clip_length_in_frames, frames_between_clips, frame_rate)
+
+    def _compute_frame_pts(self):
+        self.video_pts = []
+        self.video_fps = []
+        # TODO maybe paralellize this
+        for video_file in self.video_paths:
+            clips, fps = read_video_timestamps(video_file)
+            self.video_pts.append(torch.as_tensor(clips))
+            self.video_fps.append(fps)
+
+    def compute_clips(self, num_frames, step, frame_rate=None):
+        """
+        Compute all consecutive sequences of clips from video_pts.
+        Always returns clips of size `num_frames`, meaning that the
+        last few frames in a video can potentially be dropped.
+
+        Arguments:
+            num_frames (int): number of frames for the clip
+            step (int): distance between two clips
+            dilation (int): distance between two consecutive frames
+                in a clip
+        """
+        self.num_frames = num_frames
+        self.step = step
+        self.frame_rate = frame_rate
+        self.clips = []
+        for video_pts, fps in zip(self.video_pts, self.video_fps):
+            if frame_rate is not None:
+                # divup, == int(ceil(fps / frame_rate))
+                dilation = max(int((fps + frame_rate - 1) // frame_rate), 1)
```

Comment:
I think you should also modify `step`, otherwise the step if still relative to the original frame_rate.
unit test:
```
video_pts = torch.arange(11)
fps = 10
frame_rate = 5
num_frames = 3
step = 3  # same as num_frames, so I expect non-overlapping clips
```
I expect to get [0, 2, 4], [6, 8, 10] 
But the current code will give [0, 2, 4], [3, 5, 7], [6, 8, 10]

```
if frame_rate is not None:
    dilation = max(int((fps + frame_rate - 1) // frame_rate), 1)
    step = step * dilation
```

Reply:
Good catch!